hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSAppAttempt.java

```diff
@@ -831,7 +831,11 @@ private Resource assignContainer(
     }
 
     // The desired container won't fit here, so reserve
-    if (isReservable(capability) &&
+    // Important: Do not enable this kind of reservation in case of preemption,
+    // since the preempted application will re-reserve the preempted resources
+    // and prevent the starving application to get them
+    if (!scheduler.getConf().getPreemptionEnabled() &&
```

Comment:
This seems very bad. What happens if preemption is enabled, but no containers are preempted. However, if there are apps requesting large containers that don't fit in the free resources created by smaller containers finishing, these apps end up being starved forever.